hydromt/raster.py

```diff
@@ -560,18 +555,25 @@ def res(self) -> tuple[float, float]:
                 ddx0, ddy0 = xy1[0] - xy0[0], xy1[1] - xy0[1]
                 dx = math.hypot(x1y[0] - xy0[0], x1y[1] - xy0[1])
                 dy = math.hypot(xy1[0] - xy0[0], xy1[1] - xy0[1])
-        else:  # from coordinates
+        else:  # from coordinates; based on mean distance between cell centers
             xs, ys = self.xcoords.data, self.ycoords.data
             if xs.ndim == 1:
-                dx = xs[1] - xs[0]
-                dy = ys[1] - ys[0]
+                dxs, dys = np.diff(xs), np.diff(ys)
+                dx, dy = dxs.mean(), dys.mean()
             elif xs.ndim == 2:
-                ddx0 = xs[1, 0] - xs[0, 0]
-                ddy0 = ys[1, 0] - ys[0, 0]
-                ddx1 = xs[0, 1] - xs[0, 0]
-                ddy1 = ys[0, 1] - ys[0, 0]
-                dx = math.hypot(ddx1, ddy1)  # always positive!
-                dy = math.hypot(ddx0, ddy0)
+                dxs, dys = np.diff(xs[0, :]), np.diff(ys[:, 0])
+                ddx0 = xs[-1, 0] - xs[0, 0]
+                ddy0 = ys[-1, 0] - ys[0, 0]
+                ddx1 = xs[0, -1] - xs[0, 0]
+                ddy1 = ys[0, -1] - ys[0, 0]
+                dx = math.hypot(ddx1, ddy1) / (xs.shape[1] - 1)  # always positive!
+                dy = math.hypot(ddx0, ddy0) / (xs.shape[0] - 1)
+            # check if regular grid; not water tight for rotated grids, but close enough
+            # allow rtol 0.05% * resolution for rounding errors in geographic coords
```

Comment:
should we document this tolerance somewhere? Also the tolerance below is set using `atol` not `rtol`, is this intentional?

Reply:
This is intentional to work with resolution in projected units (metre) and geographic units degrees